File: csharp-selenium-nunit-reqnroll/SauceDemo.UiTests/Hooks/ScenarioHooks.cs
using Allure.Net.Commons;
using Reqnroll;
using Reqnroll.BoDi;
using SauceDemo.UiTests.Configuration;
using SauceDemo.UiTests.Drivers;
using SauceDemo.UiTests.Support;
using SauceDemo.UiTests.Users;

namespace SauceDemo.UiTests.Hooks;

/// <summary>
/// Per-scenario lifecycle: acquire a distinct login user, then open a browser for it, then tear both
/// down in the opposite order. Reqnroll instantiates one <see cref="ScenarioHooks"/> per scenario (via
/// its scenario-scoped BoDi container) and reuses that same instance for every [BeforeScenario] and
/// [AfterScenario] hook on it, so state set in one hook method is available to later ones on plain
/// instance fields - no ThreadLocal or static state needed, which is what keeps this safe under
/// `[assembly: Parallelizable(ParallelScope.Children)]`.
///
/// <see cref="Users.UserLease"/> and <see cref="Drivers.BrowserSession"/> are also registered into the
/// scenario's <see cref="IObjectContainer"/> (via <see cref="RegisterForStepDefinitions"/>) so step
/// definition classes can request <c>IWebDriver</c> / <c>UserAccount</c> in their own constructors for
/// free.
/// </summary>
[Binding]
public sealed class ScenarioHooks
{
    private readonly ScenarioContext _scenarioContext;
    private readonly IObjectContainer _objectContainer;

    private UserLease? _userLease;
    private BrowserSession? _browserSession;

    public ScenarioHooks(ScenarioContext scenarioContext, IObjectContainer objectContainer)
    {
        _scenarioContext = scenarioContext;
        _objectContainer = objectContainer;
    }

    [BeforeScenario(Order = 10)]
    public void AcquireUser()
    {
        var taggedUser = TagHelpers.GetTaggedUser(_scenarioContext);

        _userLease = taggedUser is not null
            ? new UserLease(taggedUser)
            : TestRunHooks.UserPool.Acquire(TimeSpan.FromSeconds(ConfigurationLoader.Settings.UserAcquireTimeoutSeconds));

        AllureApi.AddTestParameter("user", _userLease.Account.Username);
        AllureApi.AddTestParameter("worker", NUnit.Framework.TestContext.CurrentContext.WorkerId ?? "single-threaded");
    }

    [BeforeScenario(Order = 20)]
    public void CreateBrowserSession()
    {
        var settings = ConfigurationLoader.Settings;
        _browserSession = new WebDriverFactory(settings).Create();
        _browserSession.Driver.Navigate().GoToUrl(settings.BaseUrl);
    }

    [BeforeScenario(Order = 30)]
    public void RegisterForStepDefinitions()
    {
        _objectContainer.RegisterInstanceAs(_browserSession!.Driver);
        _objectContainer.RegisterInstanceAs(_userLease!.Account);
    }

    [AfterScenario(Order = 10)]
    public void CaptureFailureEvidence()
    {
        if (_scenarioContext.TestError is not null && _browserSession is not null)
        {
            AllureAttachments.AttachFailureEvidence(_browserSession.Driver);
        }
    }

    [AfterScenario(Order = 20)]
    public void QuitBrowser() => _browserSession?.Dispose();

    [AfterScenario(Order = 30)]
    public void ReleaseUser() => _userLease?.Dispose();
}

File: java-selenium-junit5-cucumber/src/test/java/com/saucedemo/uitests/hooks/ScenarioHooks.java
package com.saucedemo.uitests.hooks;

import com.saucedemo.uitests.config.ConfigurationLoader;
import com.saucedemo.uitests.config.TestSettings;
import com.saucedemo.uitests.drivers.BrowserSession;
import com.saucedemo.uitests.drivers.WebDriverFactory;
import com.saucedemo.uitests.support.AllureAttachments;
import com.saucedemo.uitests.support.AllureParameters;
import com.saucedemo.uitests.support.EnvironmentWriter;
import com.saucedemo.uitests.users.SharedUserPool;
import com.saucedemo.uitests.users.TagHelpers;
import com.saucedemo.uitests.users.UserLease;
import io.cucumber.java.After;
import io.cucumber.java.Before;
import io.cucumber.java.BeforeAll;
import io.cucumber.java.Scenario;

/**
 * Per-scenario lifecycle: lease a distinct login user, open a browser for it, then tear both down in the
 * opposite order.
 *
 * <p><strong>Hook ordering runs opposite ways for the two annotations</strong>, per Cucumber's own
 * javadoc: {@code @Before} runs <em>lower</em> order values first, while {@code @After} runs
 * <em>higher</em> values first, so teardown naturally mirrors setup. That is the reverse of Reqnroll in
 * the C# siblings (where both run ascending), and getting it backwards here would quit the browser
 * before the failure screenshot was ever taken. The numbers below are chosen so the effective order is:
 *
 * <pre>
 *   setup:    lease user (10)  ->  open browser (20)
 *   teardown: capture evidence (30)  ->  quit browser (20)  ->  release user (10)
 * </pre>
 *
 * The user is released only after its browser is fully closed, so an account is never handed to another
 * scenario while a session for it is still alive.
 */
public class ScenarioHooks {

    private final ScenarioContext context;

    public ScenarioHooks(ScenarioContext context) {
        this.context = context;
    }

    /**
     * Runs once per JVM before any scenario. Touching the settings here means a misconfigured BaseUrl
     * fails the whole run up front instead of surfacing as a confusing per-scenario failure later, and
     * writing environment.properties now means it survives a crashed or cancelled run.
     */
    @BeforeAll
    public static void beforeAll() {
        ConfigurationLoader.settings();
        EnvironmentWriter.write();
    }

    @Before(order = 10)
    public void leaseUser(Scenario scenario) {
        UserLease lease = TagHelpers.taggedUser(scenario)
                .map(UserLease::new)
                .orElseGet(() -> SharedUserPool.get().acquire());
        context.setUserLease(lease);

        // Surfacing both in the report is what makes the parallel behaviour visible: the Allure result
        // for every scenario names the account and the thread that actually ran it. Note this goes
        // through AllureParameters rather than Allure.parameter(...) - see that class for why the plain
        // shortcut silently does nothing when called from a hook.
        AllureParameters.addScenarioParameter(scenario, "user", lease.account().username());
        AllureParameters.addScenarioParameter(scenario, "worker", Thread.currentThread().getName());
    }

    @Before(order = 20)
    public void openBrowser() {
        TestSettings settings = ConfigurationLoader.settings();
        BrowserSession session = WebDriverFactory.create(settings);
        context.setBrowserSession(session);

        try {
            session.driver().get(settings.baseUrl());
        } catch (RuntimeException e) {
            // Navigation failed after the browser was already launched. The @After hooks do run for a
            // scenario whose @Before threw, so the session would still be closed - but closing it here
            // keeps the failure self-contained rather than depending on that.
            session.close();
            context.setBrowserSession(null);
            throw e;
        }
    }

    @After(order = 30)
    public void captureFailureEvidence(Scenario scenario) {
        if (scenario.isFailed() && context.hasBrowserSession()) {
            AllureAttachments.attachFailureEvidence(context.driver());
        }
    }

    @After(order = 20)
    public void quitBrowser() {
        if (context.hasBrowserSession()) {
            context.browserSession().close();
        }
    }

    @After(order = 10)
    public void releaseUser() {
        UserLease lease = context.userLease();
        if (lease != null) {
            lease.close();
        }
    }
}
